### DISCUSSION TITLE
Live Proctoring with Zoom Issue

### DISCUSSION THREAD

**Discussion by ano010** [SafeExamBrowser/seb-server#50] [Q&A] [answered]
Hi,

I have been trying to use Zoom for live-proctoring with the SEB server. It was not successful. When analyzing the logs for the Zoom marketplace app I found that the SEB server calls a Zoom API endpoint to create a meeting that does not exist in the Zoom API.

Endpoint SEB requests: https://api.zoom.us/v2/users/meetings
Actual endpoint for creating a meeting with Zoom API: https://api.zoom.us/v2/users/:userId/meetings

It seems the user id is missing when the request is made with the SEB server. is that a bug?. Can I have any clue which will be a great help to investigate further?

SEB server version: v1.2-stable-03a0e03
SEB Windows client version: 3.3.1

Thanks in advance.

I have attached the request sent by the SEB server here.
{
"endpoint": "https://api.zoom.us/v2/users/meetings",
"response_headers": [
"Set-Cookie: zm_aid=""; Domain=.zoom.us; Expires=Thu, 01-Jan-1970 00:00:10 GMT; Path=/; Secure; HttpOnly"
],
"date_time": "2021-12-17 11:22:26",
"method": "POST",
"request_body": "{"topic":"ProctoringRoom1","type":2,"start_time":"2021-12-17T05:52:26","timezone":"UTC","duration":0,"password":"******","settings":{"host_video":false,"mute_upon_entry":false,"join_before_host":true,"jbh_time":0,"use_pmi":false,"audio":"voip","waiting_room":false,"allow_multiple_devices":false}}",
"response": "N/A",
"request_headers": [
"accept-encoding: gzip,deflate",
"accept: application/json",
"authorization: ******",
"connection: close",
"content-type: application/json",
"user-agent: Apache-HttpClient/4.5.13 (Java/11.0.6)"
],
"request_params": [
],
"http_status": "405"
}

**Comment by AndreasHefti** [comment, marked as answer]
Hi

SEB Server actually uses the "https://api.zoom.us/v2/users/:userId/meetings" endpoint as you can see in the code:

![image](https://user-images.githubusercontent.com/12219424/146523516-7d3067ac-e87e-4e50-9ff4-a6eb88681e6e.png)

But SEB Server also firstly creates a user within the other endpoint "v2/users" as you also can see in the code:

![image](https://user-images.githubusercontent.com/12219424/146524407-8588f083-5ade-4c8a-b4b7-818038a97863.png)

What I can think of, without further information is, that the first Zoom API call to create a user (to manage the meeting) went wrong. Maybe because of not heaving the right privileges to do so, or due to something other unknown reason. If SEB Server is not able to detect this as an failure and try to call the Zoom API with an empty userId might actually result in calling a wrong endpoint as you've described above.

Do you have a Zoom higher plan account that is needed to use Zoom for proctoring with SEB Server?

![image](https://user-images.githubusercontent.com/12219424/146525917-af97dea4-48c7-48c3-aedf-87fd7ba737e3.png)


If yes, please also have a look at the SEB Server logs if there can something be found. You can post it here and we can have a look at it.

Best

**Reply by ano010** [reply]
Hi @AndreasHefti,

Thanks for your input. With your suggestion, I was able to integrate the Zoom service for proctoring by upgrading the Zoom plan from Basic to Pro.

**Reply by ano010** [reply]
@AndreasHefti,

I have been exploring how the SEB server uses the Zoom service. I think it would be helpful to set up an account plan in Zoom for SEB server proctoring. I have noticed the followings.

- SEB server creates a new user in the Zoom account for each break-out room.
- SEB server creates a new user in the Zoom account for Town Hall.
- SEB server creates a new user in the Zoom account for each Single Room.
- SEB server creates a licensed user in Zoom account always in the above scenarios.

I wonder why the SEB server creates a licensed user in the Zoom account? Are there any specific requirements?

One other thing, the created users are not deleted from the Zoom account even after closing the "Town Hall" or "Single Room" or "Break-Out-Room". It keeps many licensed users in the Zoom account. So the Zoom account plan can not be utilized efficiently. Sometimes a room creation fails as it reaches the maximum number of paying users in my account plan. If I am correct, we need an expensive Zoom plan with a lot of licensed users to conduct live proctoring with the SEB server. Please correct me if I am wrong. And I expect some input from you for solving the above issue?

**Reply by anhefti** [reply]
Hi,

The reason why we create a user for every meeting within the Zoom proctoring is the meeting API daily rate limit as described here: https://marketplace.zoom.us/docs/api-reference/zoom-api/methods/#operation/meetingCreate

If we would have only one user for all meetings and ad-hoc meetings for an exam, we would not be able to grantee to open as much rooms as needed during this exam without reaching the limit. So we have to create a user per meeting that managed this meeting to avoid this limit. As you can see in the code at : [ch.ethz.seb.sebserver.webservice.servicelayer.session.impl.proctoring.ZoomProctoringService.deleteAdHocMeeting](https://github.com/SafeExamBrowser/seb-server/blob/master/src/main/java/ch/ethz/seb/sebserver/webservice/servicelayer/session/impl/proctoring/ZoomProctoringService.java#L605), the user should be deleted after the meeting was deleted. If this is not the case we can consider it as a bug. In this case, please send us some more information about the problem on our side e.g. SEB Server Logs, Zoom API call errors...

Please also note that is always better to use the proctoring page's "Close" Button to close a window then just close the window with the browser window close. It still should work with the browser close but it might depend on the browser type.

**Reply by ano010** [reply]
@AndreasHefti,

Thanks for sharing the information. One more clarification. I believe, The SEB server creates a licensed user rather than a basic user since the basic user in Zoom plan has time limitations on hosting group meetings?

**Reply by anhefti** [reply]
@ano010 

>  I believe, The SEB server creates a licensed user rather than a basic user since the basic user in Zoom plan has time limitations on hosting group meetings?

Yes this is right. In the first version we'd done this with basic users and run into the time limit issue (60min) you described. Then we used licensed user to deal with all the different limitations. All this different kind of limitations that Zoom has applied to its APIs and meetings makes it some kind of difficult to deal with such proctoring scenario we have here. 

The free alternative at the moment is Jitsi Meet but with the big downside of the need of self-hosting and scaling the service what leads to a lot of extra effort.

**Reply by ano010** [reply]
@AndreasHefti 
> the user should be deleted after the meeting was deleted. If this is not the case we can consider it as a bug. In this case, please send us some more information about the problem on our side e.g. SEB Server Logs, Zoom API call errors...
> 
> Please also note that is always better to use the proctoring page's "Close" Button to close a window then just close the window with the browser window close. It still should work with the browser close but it might depend on the browser type.

It works for "Single Room Proctoring" and "Town Hall Room Proctoring". The created users are automatically deleted when closing the "Single Room" or "Town Hall". But for the "Break Out Room" It is not the case. The created users are not automatically deleted. The number of users in the Zoom account is the same as the number of Break Out Rooms even after closing all the client's connections. It is still having the same amount of users in the Zoom account even if the number of break-out rooms is decreased. I think it would be better if the users can be deleted when closing the break-out room or if the break-out room is not available or at least when closing all the client connections. This might be a bug I guess but sorry I couldn't find any logs related to this in the SEB server. Maybe it is not a bug and can be considered a new feature. Let me know if you have any different thoughts.

**Reply by anhefti** [reply]
@ano010 

Thanks for your info and investigation.
"Break Out Room" is just an overall-name of the "Single Room Proctoring" and the "Town Hall Room Proctoring" since this meetings are created ad-hoc and braking-out users from the normal "Collecting Rooms".

Please have a look at https://github.com/SafeExamBrowser/seb-server/blob/master/src/main/java/ch/ethz/seb/sebserver/webservice/servicelayer/session/impl/proctoring/ZoomProctoringService.java Line 445 in the disposeBreakOutRoom Method. At the end of this method, deleteAdHocMeeting gets been called that deletes the meeting as well as the user. 

I assume that you mean the "Collecting Rooms" or ServiceRooms in the exam monitoring instead of "Break Out Room"?
![image](https://user-images.githubusercontent.com/39078777/174019141-26433ed1-4064-42f9-acdb-5b6a572729c6.png)

The "Collecting Rooms" or ServiceRooms, that also has a separate user, are living longer, until the Exam ends or is been deleted. Then this users should also be deleted. These rooms live as long as the exam is running and even if a such room/meeting gets empty the meeting remains on Zoom side as well as on SEB Server side, ready to fill up again with SEB clients that will eventually connect to the exam.

So if you finish up (set end date) or delete your exam, the meetings and users generated for this exams collecting rooms are also been delete. You can check this if you want too, and have a look at your Zoom account users after the Exam has finished or has been deleted.

Hope this clarifies this behavior. If after an exam has ended or has been deleted and there are still Zoom account users (from SEB Server) remaining, then this can be considered as a bug.

**Reply by ano010** [reply]
@AndreasHefti,
Yes by "Break Out Room" I meant "Collecting Room". Thanks for the prompt and detailed explanation. Now it makes sense to me. I have also tried to delete the exam. It worked as expected. It deletes the Zoom users as well. I really appreciate your help.

I have one more question. It's not related to the above-mentioned issue anyway. In the SEB server proctoring UI I can only see the video of one client (Sometimes the video of the person who talks) when multiple clients join and the video is enabled. Is it possible to see all the client's videos in the frame? (It should be the way since the proctor needs to monitor all the students). Do I have to configure anything to enable it? It's the same in the client's browser as well. When we use Zoom alone we have this future. So does the SEB server restrict this feature?

**Reply by anhefti** [reply]
@ano010 No SEB Server does not restrict the gallery view, if this is what you mean. It might be that the gallery view is not the default setting within your overall zoom account but I'm not sure about that. 
Normally you are able to switch between the views within the Zoom UI by using the View action on the right upper corner of the Zoom Meeting UI: https://www.wikihow.com/Get-Gallery-View-on-Zoom

If you use the Zoom desktop-app this should work properly but we have also encountered that in some browsers the view selection on the upper right corner of the Zoom UI is just not available. It seems that this is not supported by Zoom for all browsers. So if you can, I recommend to go with the Zoom Desktop App for the collecting room proctoring:

![image](https://user-images.githubusercontent.com/39078777/174313382-8e036a4f-2963-4ab8-8ce9-4be05b4a3983.png)

**Reply by ano010** [reply]
@AndreasHefti Thanks for the information. It works with the Zoom desktop application. I have also tested this feature by joining the Zoom meeting without the SEB server using the Chrome browser. There too I could see the gallery view option. For your information, I think there is an issue with the SEB proctoring UI which hides the gallery view.

**Reply by anhefti** [reply]
@ano010 Interesting, thanks for your information. We use the usual [Zoom Meeting Web SDK](https://devsupport.zoom.us/hc/en-us/articles/360060333111-How-to-embed-Zoom-into-a-website) within SEB Server. Regarding to the Zoom Info Page here: https://marketplace.zoom.us/docs/sdk/native-sdks/web#browser-support The gallery view should be available for chromium browsers under some circumstances but not for the other browsers. I will check this as soon as I find some time. Meanwhile you can find the [Zoom Meeting HTML page here](https://github.com/SafeExamBrowser/seb-server/blob/master/src/main/resources/ch/ethz/seb/sebserver/gui/service/session/proctoring/zoomWindow.html) if you want to have a look and try some tweaks there.

**Reply by ano010** [reply]
Thanks, @AndreasHefti,

I have a question, there is an option to add an SDK key and secret when enabling proctoring in the SEB server. What is the real use case? It works with App key and secret. isn't that enough? In which scenario I should use the SDK key and secret. It would be really helpful if you can give some details.

**Reply by anhefti** [reply]
@ano010 Currently in SEB for windows we use the same [Zoom meeting web SDK](https://marketplace.zoom.us/docs/sdk/native-sdks/web) like for proctoring but for SEB Mac and iOS we use the [Zoom meeting SDK](https://marketplace.zoom.us/docs/sdk/native-sdks/introduction/). Since Zoom has different keys for one and the other, SEB Server needs to support both, APP Key and SDK Key. Maybe we will switch to the   [Zoom meeting SDK](https://marketplace.zoom.us/docs/sdk/native-sdks/introduction/) also for SEB for Windows if possible but currently, if you want to apply an Exam with different SEB clients (Windows, Mac, iOS) you need to register both keys in SEB Server.

**Reply by ano010** [reply]
@AndreasHefti I developed a sample node express server application to render the same HTML page with the same Zoom Meeting Web SDK which is used in the SEB server. I was able to configure api_key, api_secret, signature, etc., and integrate the Zoom API. I could see the gallery view option. It seems the SharedArrayBuffer is not enabled due to some missing cross-origin configuration in the SEB server which prevents the gallery view option in the SEB server. [Web SDK Gallery View](https://marketplace.zoom.us/docs/sdk/overview/websdk-gallery-view/)

**Reply by anhefti** [reply]
@ano010  This is a very useful information, thank you very much for your effort on this. The next version of SEB Server (1.4.0) is coming soon (Aug. - Sept. 22) and we are already planing and working on the following release (1.5). I will create an Issue for this within the release 1.5 and will also considering a patch for 1.4.1 if this is feasible. Will keep you informed about this.

**Reply by anhefti** [reply]
@ano010 Hi, About the CORs issue, I'm not sure if the Issue is on SEB Server side and can be fixed on SEB Server side, since SEB Server just sends a HTML page to the browser to open the Zoom Meeting with the Zoom webSDK. So for my understanding, CORs must be enabled on the Zoom server to allow requests from SEB Server domain, as far as I understand this correctly.
I also see no CORs related java-script exception in the browser-console when I open up a Zoom meeting from SEB Server (see below).

I have achieved to send the respective response headers but with no effect on Firefox:
![image](https://user-images.githubusercontent.com/39078777/187664454-3c2ebe7e-5d46-49f6-b305-bb850e7a9f1f.png)
 

If you were able to solve this issue with your own HTML page, what exactly did you do related to CORs settings for this on the server? Have you allowed just the zoom URL/domain or the SEB Server domain.

Thanks for your help, we will try to fix this issue if it is possible to fix in SEB Server side.

**Comment by ano010** [comment]
Hi @AndreasHefti ,

I was able to set up the Zoom proctoring with SEB a couple of months ago. It worked fine without any issues. But I have an issue, I could not achieve this now. I have been trying to reset all as like what I did previously. I have noticed an error message in the SEB client Zoom frame. It displays "null's server IP address could not be found".

I changed the client versions and tried with different machines as well. But the issue is still there.

Can I get an idea of what could be the cause?

I hereby attach the screenshot of the SEB client window.

![MicrosoftTeams-image](https://user-images.githubusercontent.com/33832254/164375983-28303eb2-dbee-478f-8dfb-a3d91d24ff20.png)

Thanks

**Reply by anhefti** [reply]
Hi, this might be related to wrong proctoring settings entered within the SEB Server/Exam/Proctoring Settings or due to another problem with the proctoring service. Can you please check your proctoring settings in the Exam and provide us with log information from the SEB Server (in the time when the proctoring settings are saved and when SEB clients connect to the SEB Server) and from the SEB client as well? This would be very helpful.

**Reply by ano010** [reply]
Hi, I tried to re-create and enable proctoring. I don't get any logs regarding proctoring when it is enabled or SEB client connection when it is connected. But I have a workaround now. I have been setting up the exam configuration with Zoom and Jitsi Meet Enabled. When I disabled the Jitsi Meet option the Zoom proctoring is working without any issue. But it should work with the Jitsi Meet enabled as well regardless of what the proctoring service we use as Zoom is already enabled. Is that a bug in the server?

SEB server version: v1.2-stable

**Reply by anhefti** [reply]
Hi and thank you for your info, very appreciated. 

This looks like a bug, yes. But having both (Zoom and Jitsi) enabled in a exam configuration makes not a lot of sense anyway, since there can only be one dedicated proctoring service used for a particular exam.

I will have a closer look at this issue in the next days with my colleague @dbuechel to find good fix for that (either to make the proctoring service selection in the exam configuration exclusive or to deal with multiple service config on the SEB side)

By the way, there is a newer version of SEB Server available with some new features and improvements: v1.3-stable. Please have a look at:  https://github.com/SafeExamBrowser/seb-server/releases/tag/v1.3-stable
